File: 19. Remove Nth Node From End of List .cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // Two pass: O(N)time, O(1)space, 7.14%
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = 0, cnt = 0;
        ListNode* dummy = head;
        while(head){
            len++;
            head = head->next;
        }
        // bug test case: [1,2], n=2, delete head
        if(len == n) return dummy->next;
        head = dummy;
        while(head){
            cnt++;
            if(cnt + n == len){
                head->next = head->next->next;
                break;
            }
            head = head->next;
        }
        return dummy;
    }

    // one pass: O(N)time & space, 7.14%
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int len = 0;
        unordered_map<int, ListNode*> nodes;
        ListNode* dummy = head;
        while(head){
            len++;
            nodes[len] = head;
            head = head->next;
        }
        // bug test case: [1,2], n=2, delete head
        if(len == n) return dummy->next;
        nodes[len - n]->next = nodes[len - n]->next->next;
        return dummy;
    }

    // Two pointers (one pass): O(N)time, O(1)space, 30.71%, ok
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(-1);  // bug: don't forget to initialize
        dummy->next = head;
        ListNode* slow = dummy;
        ListNode* fast = dummy;
        while(n--)  // fast pointer proceed n step first
            fast = fast->next;
        while(fast->next){  // when fast is the last elem, slow will be last n-1th
            slow = slow->next;
            fast = fast->next;
        }
        slow->next = slow->next->next;
        return dummy->next;
    }
};

File: 19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    
    # hash table: O(N) time & space
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        index, i = {}, 1
        while head:
            index[i] = head
            head = head.next
            i += 1
        if i - n == 1:   # BUG: remove first element 
            return index[2] if 2 in index else None   # corner case: one-elem list 
        else:
            index[i - n - 1].next = index[i - n + 1] if i - n + 1 in index else None 
        return index[1]
     
    # brute force: O(N) time, O(1) space
    def removeNthFromEnd(self, head, n):
        length, dummy = 0, head
        while head:
            head = head.next
            length += 1
        if length - n == 0:
            return dummy.next 
        
        head, i = dummy, 1
        while dummy and i != length - n:
            dummy = dummy.next
            i += 1
        dummy.next = dummy.next.next
        return head

    # two pointers: O(N) time, O(1) space
    def removeNthFromEnd(self, head, n):
        slow, fast, dummy = head, head, head
        while n:
            fast = fast.next
            n -= 1
        if fast:
            fast = fast.next
        else:
            return dummy.next
        while fast:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return head